src/videoencodingmonitor/VideoEncodingMonitor.ts

```diff
@@ -0,0 +1,173 @@
+  private continuousHwEncodeFailureCnt: number = 0;
+  static readonly CONTINUOUS_HIGH_ENCODE_CPU_THRESHOLD = 10;
+  static readonly CONTINUOUS_HW_ENCODE_FAILURE_THRESHOLD = 5;
+  static readonly HIGH_ENCODE_CPU_MS_THRESHOLD: number = 500;
+
+  /**
+   * @param context Take context from monitor task
+   */
+  constructor(private context: AudioVideoControllerState) {}
+
+  /**
+   * Check video encoding metrics for high CPU usage with software encoder
+   * software encoder and degrade video codec when necessary.
+   */
+  private checkHighEncodeCpu(videoMetricReport: { [id: string]: { [id: string]: {} } }): void {
+    let cpuLimitationDuration: number = 0;
+    let totalEncodingTimeInMs: number = 0;
+    let isHardwareEncoder: boolean = false;
+
+    const attendeeId = this.context.meetingSessionConfiguration.credentials.attendeeId;
+
+    const metricData: { [id: string]: { [id: string]: {} } } = videoMetricReport[attendeeId];
+    const streams = metricData ? Object.keys(metricData) : [];
+    if (streams.length === 0) {
+      return;
+    }
+
+    for (const ssrc of streams) {
+      for (const [metricName, value] of Object.entries(metricData[ssrc])) {
+        switch (metricName) {
+          case 'videoUpstreamTotalEncodeTimePerSecond':
+            totalEncodingTimeInMs = Math.trunc(Number(value));
+            break;
+          case 'videoUpstreamEncoderImplementation':
+            isHardwareEncoder = value as boolean;
+            break;
+          case 'videoUpstreamCpuQualityLimitationDurationPerSecond':
+            cpuLimitationDuration = Math.trunc(Number(value));
+            break;
+        }
+      }
+    }
+
+    const cpuUsageIsHigh =
+      !isHardwareEncoder &&
+      (totalEncodingTimeInMs >= VideoEncodingMonitor.HIGH_ENCODE_CPU_MS_THRESHOLD ||
+        cpuLimitationDuration > 0);
+    if (cpuUsageIsHigh) {
+      this.continuousHighEncodeCpuCnt++;
+      if (
+        this.continuousHighEncodeCpuCnt > VideoEncodingMonitor.CONTINUOUS_HIGH_ENCODE_CPU_THRESHOLD
+      ) {
+        this.degradeVideoCodec();
+        this.context.statsCollector.videoCodecDegradationHighEncodeCpuDidReceive();
+        this.continuousHighEncodeCpuCnt = 0;
+      }
+    } else {
+      this.continuousHighEncodeCpuCnt = 0;
+    }
+  }
+
+  /**
+   * Check video encoding metrics for hardware video encoder failures and
+   * degrade video codec when necessary.
+   */
+  private checkHwEncodeFailure(videoMetricReport: { [id: string]: { [id: string]: {} } }): void {
+    let videoInputFps: number = 0;
+    let videoEncodeFps: number = 0;
+    let isHardwareEncoder: boolean = false;
+
+    const attendeeId = this.context.meetingSessionConfiguration.credentials.attendeeId;
+
+    const metricData: { [id: string]: { [id: string]: {} } } = videoMetricReport[attendeeId];
+    const streams = metricData ? Object.keys(metricData) : [];
+    if (streams.length === 0) {
+      return;
+    }
+
+    for (const ssrc of streams) {
```

Comment:
nit: seems like maybe we could create helpers here.

Reply:
Simplified metric update with `getObservableVideoMetricValue` function